src/script/descriptor.cpp

```diff
@@ -19,6 +19,117 @@
 
 namespace {
 
+////////////////////////////////////////////////////////////////////////////
+// Checksum                                                               //
+////////////////////////////////////////////////////////////////////////////
+
+// This section implements a checksum algorithm for descriptors with the following properties:
+// * Every 1 character substitution error counts as 1 or 2 symbol errors, but:
+//   * An error substituting a character from 0123456789()[],'/*abcdefgh@:$%{} for another in
+//     that set always counts as 1 symbol error.
+//   * A case error always counts as 1 symbol error.
+// * Any 1 symbol error is always detected.
+// * Any 2 symbol error in a descriptor of up to 49143 characters is always detected.
+// * Any 3 symbol error in a descriptor of up to 501 characters is always detected.
+// * Any 4 symbol error in a descriptor of up to 93 characters is always detected.
+// * Random errors have a chance of 1 in 2**40 of being undetected.
+// * Has a slightly better chance than that for 5 symbol errors up to length 501 descriptors.
+//
+// These properties are achieved by expanding every group of 3 (non checksum) characters into
+// 4 GF(32) symbols, over which a cyclic code is defined.
+
+/*
+ * Interprets c as 8 groups of 5 bits which are the coefficients of a degree 8 polynomial over GF(32),
+ * multiplies that polynomial by x, computes its remainder modulo a generator, and adds the constant term val.
+ *
+ * This generator is G(x) = x^8 + {30}x^7 + {23}x^6 + {15}x^5 + {14}x^4 + {10}x^3 + {6}x^2 + {12}x + {9}.
+ * It is chosen to define an cyclic error detecting code which is selected by:
+ * - Starting from all BCH codes over GF(32) of degree 8 and below, which by construction guarantee detecting
+ *   3 errors in windows up to 19000 symbols.
+ * - Taking all those generators, and for degree 7 ones, extend them to degree 8 by adding all degree-1 factors.
+ * - Selecting just the set of generators that guarantee detecting 4 errors in a window of length 512.
+ * - Selecting one of those with best worst-case behavior for 5 errors in windows of length up to 512.
+ *
+ * The generator and the constants to implement it can be verified using this Sage code:
```

Comment:
@gmaxwell @sdaftuar What do you think about the inclusion of the Sage code here? I could do the same for the Bech32.

Reply:
I think that's overkill.